src/uct/tcp/tcp_listener.c

```diff
@@ -15,62 +16,69 @@ static void uct_tcp_listener_conn_req_handler(int fd, void *arg)
     char ip_port_str[UCS_SOCKADDR_STRING_LEN];
     struct sockaddr_storage client_addr;
     ucs_async_context_t *async_ctx;
-    uct_tcp_sa_arg_t *sa_arg_ctx;
+    uct_tcp_sockcm_ep_t *ep;
+    uct_ep_params_t params;
     ucs_status_t status;
     socklen_t addrlen;
-    int accept_fd;
+    int conn_fd;
 
     ucs_assert(fd == listener->listen_fd);
 
-    async_ctx = listener->super.cm->iface.worker->async;
-
     addrlen   = sizeof(struct sockaddr_storage);
     status    = ucs_socket_accept(listener->listen_fd,
                                   (struct sockaddr*)&client_addr,
-                                  &addrlen, &accept_fd);
+                                  &addrlen, &conn_fd);
     if (status != UCS_OK) {
         return;
     }
 
-    ucs_assert(accept_fd != -1);
+    ucs_assert(conn_fd != -1);
 
-    ucs_trace("server accepted a connection request from client %s",
-              ucs_sockaddr_str((struct sockaddr*)&client_addr, ip_port_str,
-                               UCS_SOCKADDR_STRING_LEN));
+    ucs_trace("server accepted a connection request (fd=%d) from client %s",
+              conn_fd, ucs_sockaddr_str((struct sockaddr*)&client_addr,
+                                        ip_port_str, UCS_SOCKADDR_STRING_LEN));
 
     /* Set the accept_fd to non-blocking mode
      * (so that send/recv won't be blocking) */
-    status = ucs_sys_fcntl_modfl(accept_fd, O_NONBLOCK, 0);
+    status = ucs_sys_fcntl_modfl(conn_fd, O_NONBLOCK, 0);
     if (status != UCS_OK) {
         goto err;
     }
 
-    sa_arg_ctx = ucs_malloc(sizeof(uct_tcp_sa_arg_t), "accept sa_arg_ctx");
-    if (sa_arg_ctx == NULL) {
-        ucs_error("failed to allocate memory for a listen_ctx");
+    params.field_mask        = UCT_EP_PARAM_FIELD_CM                |
+                               UCT_EP_PARAM_FIELD_CONN_REQUEST      |
+                               UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS ;
+    params.sockaddr_cb_flags = UCT_CB_FLAG_ASYNC;
+    params.cm                = listener->super.cm;
+
+    status = UCS_CLASS_NEW(uct_tcp_sockcm_ep_t, &ep, &params);
```

Comment:
maybe `uct_ep_create`?
`UCT_EP_PARAM_FIELD_CONN_REQUEST` is set in field mask but field is not initialized

Reply:
I dont think it's a good practice to call a uct API function from inside uct..

Reply:
then maybe `cm->ops.ep_create`?

Reply:
why not with the ctor?
(also, in the next pr i plan to change the function for `cm->ops.ep_create` to a wrapper that would call the ctor so that the wrapper would be used from the api call.)

Reply:
potentially `cm->ops.ep_create` might do more things than ctor, for example, wrapper which calls ctor + some additional logic... but since it's inside the same module, I'm OK with ctor as well.

Reply:
I see the reason of ctor call... API call does not call ctor at all, it initializes internally pre-created EP  and pass it to user, right?

Reply:
exactly.
(will be done in future prs...)